### ISSUE TITLE
Frequent disconnects while busy running in Google Colab

### ISSUE THREAD

**Issue by cwreynolds** [googlecolab/colabtools#2965]
In the last week or two I have seen frequent disconnects while trying to run a lengthy training run. A month or two ago this notebook seemed to be working pretty reliably. My code has definitely changed but those internal details seem like they should be unrelated to the operation of Colab.

**Describe the current behavior**
During a run of only a few minutes, I will see several pop-up (actually “slide up from the bottom”) notifications saying “Runtime disconnected.” When this happens it disrupts the Resources displays (sidebar says “not connected to runtime”) which makes it hard to track resource usage.

As described in a [question with the same title on stackoverflow](https://stackoverflow.com/questions/73179443/frequent-disconnects-while-busy-running-in-google-colab) my actual training seems to permanently disconnect about an hour into the multi-hour run. I am not yet able to reproduce that in a clean test case, but the frequent “faux disconnects” are easy to reproduce. See this Colab notebook [DisconnectTest](https://colab.research.google.com/drive/19uFqBsjXsarSMrMsxRjG3Lq4SL1ZEOsR) and this [screen grab video](https://www.youtube.com/watch?v=RTU0o6_onnM) showing a disconnect at 1:36.

**Describe the expected behavior**
There would **not** be frequent disconnects. The Resources displays would **not** be disrupted. (And my hours-long training run would continue until completion.)

**What web browser you are using**
Chrome — Version 103.0.5060.134 (Official Build) (arm64)
On macOS Monterey 12.4 Apple M1 Max

**Additional context**
Test case notebook: https://colab.research.google.com/drive/19uFqBsjXsarSMrMsxRjG3Lq4SL1ZEOsR 
Video of sessions with issue: https://www.youtube.com/watch?v=RTU0o6_onnM
Question on stack overflow: https://stackoverflow.com/q/73179443/1991373

**Comment by craigcitro** [comment]
Thanks for the excellent description!

It's hard to diagnose connection issues via Github -- could I ask you to `Help` -> `Send feedback` the next time this happens and reference this github issue?

Thanks!

**Comment by cwreynolds** [comment]
Thanks @craigcitro for your reply.

I ran the sample notebook (https://colab.research.google.com/drive/19uFqBsjXsarSMrMsxRjG3Lq4SL1ZEOsR ) waited for the first disconnect, then did a Send Feedback. The comment mentions this Issue by name and URL.

**Comment by cperry-goog** [comment]
Tracked internally at b/241479492

**Comment by blois** [comment]
I'm curious if the issue continues to occur in incognito mode? It could be caused by a browser extension.

**Comment by cwreynolds** [comment]
@blois thank you for looking at this.

**Short answer: running in Incognito mode does not appear to change the symptom.**

I ran both my original (research related) notebook and the [minimal test case notebook](https://colab.research.google.com/drive/19uFqBsjXsarSMrMsxRjG3Lq4SL1ZEOsR) in Incognito mode. They both quickly exhibited the momentary disconnects. In the test case notebook, it happened about 82 seconds into the run, epoch ~20 of 50. As before, this disrupts the Resources sidebar — causing it to say “not connected to runtime” — until the run terminates and the sidebar refreshes.

**Comment by blois** [comment]
Does this occur for a simple non-GPU notebook, such as:

```python
import time
time.sleep(100)
```

**Comment by cwreynolds** [comment]
@blois: yes it does! Good suggestions for zeroing in on the scope of the issue.

The first time it was just a few seconds into the 100 second run. So I squeezed the window down and set up to record that part of the screen: this time it ran for 1:12 (72 of 100 seconds). I edited out the waiting time, but here you see the disconnect message, the Resources sidebar lose connection, and the runtime recovers and goes into Busy state:

https://youtu.be/pOUpwVqZH_Q

**Comment by cwreynolds** [comment]
**Aha!**

I came back here to thank the people who looked at this issue (@blois, @cperry-goog, @craigcitro) and to sort of say goodbye. I have since moved to running my experiments locally on my Apple M1 MacBook Pro.

The “just sleep for 100 seconds” example above was so trivial that I decided to try it again and verify the problem still exists. I opened a new Colab notebook, copied in that text, and ran the cell. It did nothing (which is good!) for 100 seconds then exited. Hmm. I opened the Resources sidebar and confirmed with another run that it stayed connected. Huzzah!

Then I realized I was **running Colab from my Safari browser**, whereas before I had been using Chrome. OK, I tried to run the test again using **Colab on Chrome. And it failed,** like before, running only a few seconds before a quick disconnect, causing the Resources sidebar to go blank. I repeated the test and got the same result.

Only four quick test runs, but this strongly **suggests the problem is some interaction between Colab and Chrome** (and perhaps macOS Monterey (Version 12.5.1), and perhaps M1 Apple Silicon). Yet Colab seems to work fine on M1/Monterey/Safari.

**The disconnects today on Chrome were with “Version 104.0.5112.101 (Official Build) (arm64)”** while the previous ones were with “Version 105.0.5195.102 (Official Build) (arm64)”.

I hope this helps narrow down the symptoms of the problem I reported last month, and allows it to be reproduced by someone else with an M1/arm64/Apple Silicon machine.

**Comment by blois** [comment]
If it repro's in Chrome but not Safari another potential issue is a browser extension causing the problem. Trying it in an incognito window often disables extensions and may avoid reproducing the issue.

**Comment by cwreynolds** [comment]
Yes, you suggested that [before](https://github.com/googlecolab/colabtools/issues/2965#issuecomment-1221191215) and the problem still happened in Incognito. But for the sake of Science! I tried it again, with the `time.sleep(100)` benchmark. Still failed in Chrome Incognito.

**Comment by Marin-Kitagawa** [comment]
I have the same problem as well. But the timing is not that short. While it is executing (primarily concerted with AI & ML algorithms), it just randomly disconnects. I think the time was less than 300 seconds. It is hard to work with ML in Colab.

**Comment by shuvayan** [comment]
I am facing this issue currently and not even a single epoch is completing if the session is left idle. Since colab is built mainly for notebook like use cases , I think multi-hour runs are not supported.

**Comment by cperry-goog** [comment]
Multi-hour runs are supported - do you see an error message or have more info we can use to debug?

**Comment by cwreynolds** [comment]
Hi @Marin-Kitagawa, @shuvayan, and @cperry-goog. It has been two years since I was using colab for long training runs, but I did find a work-around: when I switched to a non-Chrome browser (I use Safari) the problem immediately went away. I could reliably make multi-hour runs without problems See: https://github.com/googlecolab/colabtools/issues/2032#issuecomment-1292202803

But a lot will have changed in two years, so my experience may not be relevant to current versions.

**Comment by shuvayan** [comment]
There is no error message, it just gets disconnected after sometime , if the screen is locked. I just need to stay awake so that the screen doesnt get locked and the code continues to run.

**Comment by cwreynolds** [comment]
@cperry-goog there is a minimal failing case (proposed by @blois in 2022) and some discussion above in two comments:

https://github.com/googlecolab/colabtools/issues/2965#issuecomment-1221425337
https://github.com/googlecolab/colabtools/issues/2965#issuecomment-1237570020

**Comment by RobertoVzz** [comment]
Dealing with a very similar issue, getting runtime disconnects at random points of notebook executions.
Also working with AI/ML, and it only happens to me when running on sessions with GPU (T4/L4).
Running on Safari didn't work for me. Did "last longer" than on Chrome attempts, though.

**Comment by spencersgoogle** [comment]
@RobertoVzz this is difficult to diagnose from Github. Please file in-product feedback at Help -> Send Feedback and include a link to this GitHub issue. Please make sure to share an accessible link to the notebook in question.

**Comment by 0wwafa** [comment]
It's happening more and more lateley... even a sleep(1000) disconnects every minute or so even if I am moving the mouse and I am on the colab page!

On the scratchpad it happens even more!

**Comment by 0wwafa** [comment]
> [@RobertoVzz](https://github.com/RobertoVzz) this is difficult to diagnose from Github. Please file in-product feedback at Help -> Send Feedback and include a link to this GitHub issue. Please make sure to share an accessible link to the notebook in question.

it's not a particular notebook... open the scratchpad, connect to the basic VM... it's happening continuously!

**Comment by pjmartel** [comment]
I have the same problem. I am on the Pro subscription of Colab, and I get frequent disconnects when working interactively on notebooks. They happen when I run a cell, and more frequently when running import commands. I often teach a class with Colab (where I am the only Pro user), and it is just my notebook that keeps disconnecting. This is terribly annoying. Please give me some direction on how to find the cause of this issue.
This happens with scratchpad as well as any other notebooks, and the disconnects happen right after a start my session, so it is not a problem of resource usage.
